reps/2023-12-04-accelerated-dag.md

```diff
@@ -0,0 +1,360 @@
+worker death, we will propagate the error to the DAG outputs.
+
+Currently, Ray Core also provides automatic task-level re-execution. At
+the moment, we do not plan to support this, as it would likely require
+extra overhead from tracking intermediate task outputs. However,
+automatic DAG-level re-execution may be possible.
+
+## Application-defined transports
+
+***Goal:** Give the application greater control over the communication
+method, but allow Ray Core to schedule the communication.* This could
+allow us to keep current Ray Core system optimizations such as:
+
+1.  Overlapping communication with task execution.
+2.  Handle communication failures gracefully through retries.
+3.  Control communication concurrency / resource usage.
+
+Accelerated DAGs will ship with a default transport for objects passed
+along the DAG's edges. This will use shared memory if the
+sender/receiver are colocated, and shared memory/Ray Core for
+cross-node.
+
+To support specialized transports, we will extend the DAG API to allow
+application-defined transports that can be called from either Python or
+C++. The exact API is TBD, but here is a simple initial proposal based
+on a Channel concept:
+
+```python
+class Transport:
+  def __init__(self, sender: ray.ActorHandle, receiver: ray.ActorHandle):
+    pass
+  def send(self, send_meta: BufferMetadata):
+    """Ideally async. Called by the sender"""
+    pass
+  def recv(self, recv_meta: BufferMetadata) -> Buffer:
+    """Ideally async. Called by the receiver."""
+    pass
+```
+
+When an application-defined transport is used, we will use the default
+transport to synchronize between the sender and receiver, but we only
+send the BufferMetadata in along the default transport. This will act as
+a signal to begin the actual \`Transport.send\` and \`Transport.recv\`
+of the application data.
+
+Initially, our goal is to add a UCX-based transport, which supports
+GPU-GPU and RDMA, among others. For GPU objects, we can also provide a
+fallback transport, i.e. when a specialized GPU-GPU transport is not
+available, the fallback transport copies the GPU object to the Ray
+object store, then uses the default transport to transfer the object.
+
+The other potential complexity in this work is supporting collective
+communication, which can deadlock if not properly scheduled. During the
+initial prototyping, we can continue to execute collective communication
+out-of-band, as part of the application code. However, to realize the
+long-term goals, we will introduce an API that captures collective
+communication ops as part of the DAG. An example API:
+
+```python
+workers = [Worker.remote() for worker in workers]
+# A task called on the CollectiveGroup will be gang-scheduled on each worker.
+pool = CollectiveGroup(workers)
+outputs = pool.allreduce.bind(inputs)
+```
+
+Workloads
+=========
+
+Possible validation workloads fall under two different categories:
```

Comment:
It would be great to show the code for a few example workloads, for example, the scatter-gather vLLM example, and perhaps something more complex like Gpipe parallelism.